LSTM SymInt-aware changes & meta registration (non-cuDNN CUDA)

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90980
* #90944
* __->__ #90701
* #90702
* #90714
* #90697

Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some [nasty stuff](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp#L1567) with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)

jbschlosser committed:
LSTM SymInt-aware initial work (incomplete)

[ghstack-poisoned]

jbschlosser committed:
Update on "LSTM SymInt-aware initial work (incomplete)"


Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some nasty stuff with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)


[ghstack-poisoned]

jbschlosser committed:
Update on "LSTM SymInt-aware initial work (incomplete)"


Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some nasty stuff with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)


[ghstack-poisoned]

jbschlosser committed:
Update on "LSTM SymInt-aware initial work (incomplete)"


Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some nasty stuff with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)


[ghstack-poisoned]

ezyang commented:
btw, the thnn stuff is independently useful, and I think unblocks some models

ezyang reviewed (commented):
need tests for the new metas

jbschlosser committed:
Update on "LSTM SymInt-aware initial work (incomplete)"


Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some [nasty stuff](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp#L1567) with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)


[ghstack-poisoned]

jbschlosser commented:
> need tests for the new metas

Working on tests now! Since there's no OpInfo for LSTM, we can either do a one-off test or take advantage of the existing ModuleInfo entries to generate a new set of tests.

> btw, the thnn stuff is independently useful, and I think unblocks some models

I'll split this into separate PRs for thnn / cuDNN

jbschlosser committed:
Update on "LSTM SymInt-aware initial work (incomplete)"


Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some [nasty stuff](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp#L1567) with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)


[ghstack-poisoned]

ezyang commented on test/test_fake_tensor.py:
```diff
@@ -370,6 +371,36 @@ def fn(
+
+                self.assertEqual(output.shape, (L, N, D * H_out))
+                self.assertEqual(h_n.shape, (D * num_layers, N, H_out))
+                self.assertEqual(c_n.shape, (D * num_layers, N, hidden_size))
```
One downside for this test is it doesn't exercise symbolic shapes. But I guess this doesn't actually work yet?

jbschlosser replied:
That's true - the test should ideally be at the AOTAutograd level to exercise symbolic shapes, but that doesn't work yet due to #90500.

The test here just covers the meta registration for now. Tests that will exercise symbolic shapes are added in #90980 (currently with xfails for the RNNs).

jbschlosser committed:
Update on "LSTM SymInt-aware changes & meta registration (non-cuDNN CUDA)"


Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some [nasty stuff](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp#L1567) with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)


[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
